[Benjamin Tan] iP

bnjmnt4n opened the pull request:
# Duke

> “Wow, yet another semi-useful chatbot/task management service. :disappointed: ” – me ([source](https://github.com/nus-cs2103-AY2122S1/ip/pull/420#issue-719240500))

Duke is a task management service with an interactive chatbot interface. It's,

- local-first
- text-based
- semi-useful
- easy to learn
- written in Java, ~~sadly~~

All you need to do is,

1. download it from [here](https://github.com/bnjmnt4n/ip/releases/tag/A-Jar).
1. execute via `java -jar duke.jar`
1. ~~tasks will get completed for you~~ you still have to do tasks yourself, but maybe this helps :thinking: 

And it is **FREE**!

Features:

- [X] Managing tasks
- [X] Managing deadlines
- [ ] Reminders (coming soon *(hopefully)*)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

*Note:* feel free to ignore the `.nix` files, they are used for dependency management for my system.

Commit:
Add Gradle support

bnjmnt4n committed:
Add Nix flake files

bnjmnt4n committed:
Add `.gitignore`

bnjmnt4n committed:
Add basic greet, echo, and delete functionality

bnjmnt4n committed:
Store text input by users and add `list` command

bnjmnt4n committed:
Add `jdt-language-server`

bnjmnt4n committed:
Improve formatting

bnjmnt4n committed:
Add `Task` implementation

bnjmnt4n committed:
Duke: make use of `Task`

bnjmnt4n committed:
Add `done` command to mark task as completed

bnjmnt4n committed:
Add `Event`, `Todo` and `Deadline` classes

These classes inherit from `Task`.

bnjmnt4n committed:
Deadline: rename `deadline` instance variable to `time`

bnjmnt4n committed:
Add `event`, `todo` and `deadline` commands

bnjmnt4n committed:
Setup basic text UI testing

bnjmnt4n committed:
Cleanup `.gitignore`

bnjmnt4n committed:
Duke: Convert to instantiable class

bnjmnt4n committed:
Handle unrecognized commands by throwing an exception

bnjmnt4n committed:
Add basic error handling for task creation

bnjmnt4n committed:
Task: convert to abstract class

bnjmnt4n committed:
Handle more exceptions

bnjmnt4n committed:
Add `delete` command

bnjmnt4n committed:
Duke: use enums to represent commands

bnjmnt4n committed:
Save tasks to local file

Tasks are saved to `data/duke.txt`.

bnjmnt4n committed:
Duke: remove `shutdown` instance method

Since the scanner is created externally, it should not be the
responsibility of `Duke` to close the scanner.

bnjmnt4n committed:
Tasks: make constructors accept an `isCompleted` boolean

bnjmnt4n committed:
Read tasks from file on startup

bnjmnt4n committed:
Merge branch 'branch-Level-7'

bnjmnt4n committed:
Events and deadlines: parse input text into datetimes

The time for events and deadlines is parsed using the follow format:
`YYYY-MM-DD HH:MM:SS`.

bnjmnt4n committed:
Add helper `DateTime` utility class

This class performs datetime formatting and parsing for both the `Event`
and `Deadline` classes.

bnjmnt4n committed:
Merge branch 'branch-Level-8'

bnjmnt4n committed:
Extract `Storage` class

bnjmnt4n committed:
Extract `TaskList` class

bnjmnt4n committed:
Extract `Ui` class

bnjmnt4n committed:
Extract `Command` classes

bnjmnt4n committed:
Extract `Parser` class

bnjmnt4n committed:
Fix display of completed tasks when loaded from storage

This fixed a bug due to comparison of strings using the `==` operator
instead of `Object.equal`.

bnjmnt4n committed:
Organize classes into packages

For now there are the following packages:
- duke
- duke.task
- duke.command

bnjmnt4n committed:
Update `build.gradle`

Reflect new directory structure to avoid missing classpath.

bnjmnt4n committed:
Merge branch 'add-gradle-support'

bnjmnt4n committed:
Update `.gitignore`

bnjmnt4n committed:
Use more conventional directory structure

As per the JUnit tutorial at
https://se-education.org/guides/tutorials/junit.html, use the
`src/main/java` directory structure instead of just `src`.

bnjmnt4n committed:
Add GitHub action workflow for building JARs

bnjmnt4n committed:
Add checkstyle configuration

The configuration was obtained from
https://github.com/se-edu/addressbook-level3/tree/e6c2dc4699e4da1720fbe4c2a66be8390a24cacd/config/checkstyle.

bnjmnt4n committed:
Add tests for `TaskList`

bnjmnt4n committed:
TaskList: Add `filterTasks` method

bnjmnt4n committed:
Add `find` command to filter tasks

bnjmnt4n committed:
Address code style violations

bnjmnt4n committed:
Add JavaDocs

bnjmnt4n committed:
Merge branch 'branch-A-CodingStandard'

bnjmnt4n committed:
Merge branch 'branch-A-JavaDoc'

bnjmnt4n committed:
Merge branch 'branch-Level-9'

bnjmnt4n committed:
Fix remaining code style violations

botr99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+     * @param args The command-line arguments.
+     */
+    public static void main(String[] args) {
+        Scanner input = new Scanner(System.in);
```
Perhaps a more intuitive variable name here? `input` seems to be more suited for a `String`

GraceWang2322 commented on src/main/java/duke/DateTime.java:
```diff
@@ -0,0 +1,26 @@
+     * Parses a string into a date.
+     * @param str The string containing a date.
+     */
+    public static LocalDateTime parse(String str) throws Exception {
```
maybe change the variable name str to something more specific?

GraceWang2322 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+    public void run() {
+        ui.printGreeting();
+
+        boolean shouldExit = false;
```
maybe change the boolean name to isExit to sound more like a boolean?

bnjmnt4n replied:
I think `shouldExit` is indicative that it is a boolean, and I find it clearer than `isExit`, so I left it as such.

botr99 commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+     * @param ui The instance of the {@link Ui} class.
+     * @param storage The instance of the {@link Storage} class.
+     */
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws Exception {
```
Perhaps consider adding an `@Override` to make it clearer that the subclass is implementing the abstract method of the superclass?

GraceWang2322 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Parses the user's input string into a command.
+     * @param fullCommand The full command string input by the user.
+     * @throws An exception if the input is not a valid command.
```
maybe add wht the method is returning in the header comment??

botr99 commented on src/main/java/duke/command/FilterTasksCommand.java:
```diff
@@ -0,0 +1,56 @@
+
+            for (int i = 0; i < numTasks; i++) {
+                Task item = filteredTaskList.getTask(i);
+                builder.append(i + 1);
+                builder.append(". ");
+                builder.append(item.toString());
+                if (i < numTasks - 1) {
+                    builder.append("\n");
+                }
```
Perhaps consider chaining the `append` calls together?

GraceWang2322 commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+     * Adds a deadline to the user's tasklist.
+     * @param taskList The tasklist.
+     * @param ui The instance of the {@link Ui} class.
+     * @param storage The instance of the {@link Storage} class.
```
maybe can add in wht exception the method is throwing in the header comment??

botr99 commented on src/main/java/duke/DateTime.java:
```diff
@@ -0,0 +1,26 @@
+
+public class DateTime {
+    /**
+     * Parses a string into a date.
+     * @param str The string containing a date.
```
Perhaps consider leaving a blank line in between?

GraceWang2322 reviewed (commented):
The coding standards are well-followed. The code is really clean as well! 👍 just a few nits to review

botr99 reviewed (commented):
LGTM. I like how clean your code is!

bnjmnt4n committed:
Bump `jdt-language-server`

bnjmnt4n committed:
Address PR feedback

- Make use of `@Override` when implementing abstract methods
- Some JavaDoc cleanup

bnjmnt4n commented:
Thanks for the feedback, have addressed most of it.

bnjmnt4n committed:
Gradle: install JavaFX dependencies

bnjmnt4n committed:
Add initial JavaFX application

nicolej2122 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,45 @@
+public class TaskList {
+    private List<Task> tasks;
+
+    public TaskList(List<Task> tasks) {
```
Could write descriptive header comments for all public methods.

nicolej2122 reviewed (commented):
Looks good! :) Except some public methods/ classes lack header comments

bnjmnt4n committed:
Migrate `Ui` to `ConsoleUi`

`ConsoleUi` now inherits from abstract class `Ui`. This allows creation
of a new class with JavaFX interaction inheriting from `Ui`, allowing
both GUI/text-mode interfaces for Duke.

bnjmnt4n committed:
Add JavaFX application

Based largely on
https://se-education.org/guides/tutorials/javaFxPart4.html.

bnjmnt4n committed:
Allow running duke from command line as well

bnjmnt4n committed:
Fix style violations

bnjmnt4n committed:
Merge branch 'branch-Level-10'

bnjmnt4n committed:
Add assertions to codebase

Adds two assertions to ensure that the `Duke` object in and `Ui` objects
are non-null.

bnjmnt4n committed:
TaskList: Make use of Java streams

Use Java streams to filter tasks in `TaskList`.

bnjmnt4n committed:
Merge pull request #3 from bnjmnt4n/branch-A-Assertions

Add assertions to codebase

bnjmnt4n committed:
Add `DukeException` and `InvalidCommandException`

`InvalidCommandException` represents all exceptions relating to invalid
command input.

bnjmnt4n committed:
Shift deadline input parsing into `AddDeadlineCommand`

bnjmnt4n committed:
Shift event input parsing into `AddEventCommand`

bnjmnt4n committed:
Shift todo input parsing into `AddTodoCommand`

bnjmnt4n committed:
Improve `DateTime` abstraction and add `InvalidDateTimeException`

bnjmnt4n committed:
DukeException: convert to Unix line-endings

bnjmnt4n committed:
Make use of `InvalidCommandException` in remaining commands

bnjmnt4n committed:
Add `StorageException` and simplify `Storage` class

Attempt to apply SLAP by abstracting out a common `loadTask`
abstraction.

bnjmnt4n committed:
Convert more methods to use `DukeException`

This commit also introduces `InvalidTaskException` which is thrown when
an incorrect index is used to access a task.

bnjmnt4n committed:
Throw `InvalidCommandException` during command creation

Instead of throwing `InvalidCommandException` during the execution of
the command, throw it only during command creation.

bnjmnt4n committed:
Restructure different UI modes into `duke.ui` packages

bnjmnt4n committed:
Fix tests

bnjmnt4n committed:
Merge branch 'master' into branch-A-Streams

bnjmnt4n committed:
Merge pull request #2 from bnjmnt4n/branch-A-Streams

TaskList: Make use of Java streams

bnjmnt4n committed:
Make suggestions as recommended by IntelliJ

bnjmnt4n committed:
Fix code style violations

bnjmnt4n committed:
Merge branch 'master' into branch-A-CodeQuality

bnjmnt4n committed:
Merge pull request #4 from bnjmnt4n/branch-A-CodeQuality

Improve code quality

bnjmnt4n committed:
Add initial `schedule` command

bnjmnt4n committed:
Improve error messages and comments

bnjmnt4n committed:
Fix style

bnjmnt4n committed:
Merge pull request #5 from bnjmnt4n/schedule

Add `schedule` command

bnjmnt4n committed:
Update text UI tests

bnjmnt4n committed:
CI: Run text UI tests

bnjmnt4n committed:
Improve GUI styling

bnjmnt4n committed:
Fix text UI tests

bnjmnt4n committed:
Add documentation and screenshot

bnjmnt4n committed:
Docs: minor heading addition

bnjmnt4n committed:
Fix JAR file
